### PR TITLE
Add time-rs as optional time backend

### PR DESCRIPTION
**Pull request by tyhi**
This is a preference change (and a breaking one at that) and if you would like to stay on chrono this pr can be closed. 

Performance impact on running cargo bench seems to be about 13%(+-3%) slower on generating and about -10% faster on parsing. While reducing the number of dependencies pulled in (up to 7 less on linux). Benchmarks were ran on Azure using a  Standard D8as_v4 (8 vcpus, 32 GiB memory) vm on Centos 8.4,

The datetime function uses an expect, however chrono uses an expect internally so behavior is exactly the same.

### CODE REVIEW

**Comment by huxi** [comment]
Just a quick feedback since I don't have the time right now: I'm not ignoring you!

I wouldn't merge the PR as it is but we could try to do this in a way that works with either `time` or `chrono`. That way it wouldn't necessarily have to be a breaking change.

**Comment by tyhi** [comment]
I agree, and should've done that from the start.

I've renamed the PR and change to use features to keep chrono as the default and keeping all function signatures the same. I've added a `compile_error!` to error on enabling both features though not sure if there's a better way to handle this. 

Let me know if you find anything that needs to be changed, thanks!

**Comment by huxi** [comment]
Sorry for ghosting you again. Life is kind of hard right now...

I'd prefer if you could change the code so it compiles even if both `chrono` and `time` are present at the same time.

There's a [justfile](https://github.com/huxi/rusty_ulid/blob/master/justfile) that builds the different combinations and I think it should be possible to enable all optional dependencies. It would be nice if you could extend it and make sure that `just all` still works.

File: src/lib.rs [hunk 1]
```diff
@@ -283,7 +299,7 @@ impl Ulid {
     /// # Panics
     ///
     /// Panics if called after `+10889-08-02T05:31:50.655Z`.
-    #[cfg(all(feature = "rand", feature = "chrono"))]
+    #[cfg(all(feature = "rand", feature = "time"))]
```
Comment (huxi) [inline src/lib.rs]: Still missing the `any`

File: src/lib.rs [hunk 2]
```diff
@@ -313,7 +329,7 @@ impl Ulid {
...
```
Comment (huxi) [inline src/lib.rs]: Still missing the `any`

File: src/lib.rs [hunk 3]
```diff
@@ -707,7 +723,6 @@ impl Ulid {
     /// # Ok::<(), rusty_ulid::DecodingError>(())
     /// ```
     #[cfg(feature = "chrono")]
-    #[must_use]
```
Comment (huxi) [inline src/lib.rs]: Why was this removed?

File: src/lib.rs [hunk 4]
```diff
@@ -716,6 +731,26 @@ impl Ulid {
...
+    /// use std::str::FromStr;
+    ///
+    /// let ulid = Ulid::from_str("01CAH7NXGRDJNE9B1NY7PQGYV7")?;
+    /// let datetime = ulid.datetime();
+    ///
+    /// assert_eq!(datetime.to_string(), "2018-04-07 23:39:50.168 +00:00:00");
+    /// # Ok::<(), rusty_ulid::DecodingError>(())
+    /// ```
+    #[cfg(feature = "time")]
+    pub fn datetime(&self) -> OffsetDateTime {
```
Comment (huxi) [inline src/lib.rs]: Similarly, I'd suggest to have two different methods here. One that returns the chrono type (as before) and one that returns the time type. I'm just struggling how to name it... offsetdatetime?

File: src/lib.rs (hunk 1 above)
Comment (tyhi) [reply src/lib.rs]: Fixed

File: src/lib.rs (hunk 2 above)
Comment (tyhi) [reply src/lib.rs]: Fixed

File: src/lib.rs (hunk 3 above)
Comment (tyhi) [reply src/lib.rs]: Not super sure how that got removed, re-added.

File: src/lib.rs (hunk 4 above)
Comment (tyhi) [reply src/lib.rs]: offsetdatetime makes sense and keeps it simple, works for me.

**Comment by tyhi** [comment]
No worries on the delay, definitely understandable.

So with latest commit, I performed all the changes, along with changing readme to ignore the code block (for cargo test) and saying that if both features are enabled chrono will be used over time.

I edited the justfile however I'm not super familiar with it so if it needs to be changed let me know. 

Thanks for reviewing the PR so far though!

**Comment by huxi** [comment]
I just published 1.0.0 including your changes. Had to fix a minor issue in case of both chrono and time but thanks a lot for your contribution!

**Comment by tyhi** [comment]
> I just published 1.0.0 including your changes. Had to fix a minor issue in case of both chrono and time but thanks a lot for your contribution!

Thank you for getting this merged and the assistance you provided me!
